test: dtest: move audit_test.py to test.py

andrzej-jackowski-scylladb opened the pull request:
Copied the entire audit_test.py from scylladb/scylla-dtest, to remove the entire file from scylla-dtest after this patch series is merged. The motivation is to move entire audit testing to from dtests, to make it easier to maintain and more reliable.

After audit_test.py was moved from dtests to test.py, someissues that require fixing arose due to differences between the frameworks.

No backport, moving audit_test.py to test.py is a new testing effort.

nyh commented:
We can't continue to "move" tests from dtest to test.py without taking actual steps to remove the original tests from dtest.
If we wait with this removal, nobody will remember what was actually copied and what was not.

andrzej-jackowski-scylladb commented:
> We can't continue to "move" tests from dtest to test.py without taking actual steps to remove the original tests from dtest. If we wait with this removal, nobody will remember what was actually copied and what was not.

Yes, I think you have a valid point. Our original idea was to first add the new tests to scylladb and then remove it from dtest (to avoid a moment when the test is not running in any demo) but it seems the order you propose is better. CC @piodul

nuivall commented:
> We can't continue to "move" tests from dtest to test.py without taking actual steps to remove the original tests from dtest. If we wait with this removal, nobody will remember what was actually copied and what was not.

Since we're moving the whole file, all tests (under the same name) once we merge it we can simply delete the file from dtests. I think it's safer approach than first deleting.

nyh commented:
My comment wasn't about the order, but about the plan and the technical debt. We already moved most but NOT all of the Alternator dtests to test.py, but did not (as far as I know) remove any dtest. Now this patch copies more, but again doesn't even discuss what can be removed (I didn't know you migrated *all* of the tests, I assumed some were non-gating or something), let alone mention a dtest pr which will remove the dtest.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17341/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17341/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17341/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/dtest/audit_test
#### Failed Tests (1/1620):
- [test_permissions[AuditBackendTable].dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17341/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/test_permissions_AuditBackendTable__dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_permissions[AuditBackendTable].dev.3)
#### Build Details:
- Duration: 1 hr 19 min
- Builder: spider6.cloudius-systems.com

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+            for query in query_sequence:
+                session.execute(query)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3236")
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove `require` markers. We don't need them.

And, btw, @andrzej-jackowski-scylladb suggested to close scylladb/scylla-enterprise#3236 after merge of https://github.com/scylladb/scylla-dtest/pull/5649

scylladb/scylla-enterprise#3590 is merged.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+            for query in query_sequence:
+                session.execute(query)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        with helper_class() as helper:
+            self.verify_keyspace(audit_settings=AuditTester.audit_default_settings, helper=helper)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        with helper_class() as helper:
+            self.verify_keyspace(audit_settings={"audit": "table", "audit_categories": "ADMIN,AUTH,DML,DDL,DCL", "audit_keyspaces": "ks", "extra_parameter": "new"}, helper=helper)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        with helper_class() as helper:
+            self.verify_keyspace(audit_settings={"audit": "table", "audit_categories": "ADMIN,AUTH,QUERY,DML,DDL,DCL", "audit_keyspaces": "a,b,c,ks"}, helper=helper)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        self.execute_and_validate_audit_entry(session, query=self.AUDIT_LOG_QUERY, category="QUERY", ks="audit", table="audit_log", audit_settings=audit_settings)
+
+    @pytest.mark.single_node
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        self.verify_table(audit_settings={"audit": "table", "audit_categories": "AUTH,QUERY,DDL"}, table_prefix="test_audit_categories_part1", overwrite_audit_tables=True)
+
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+            self.verify_table(audit_settings={"audit": "table", "audit_categories": "DDL, ADMIN,AUTH,DCL", "audit_keyspaces": "ks"}, helper=helper, table_prefix="test_audit_categories_part2")
+
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+
+    PasswordMaskingCase = namedtuple("PasswordMaskingCase", ["name", "password", "new_password"])
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+        assert rows_with_error[0].error is True
+        assert rows_with_error[0].consistency == "THREE"
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+                expected_error=Unauthorized,
+            )
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Please, remove this `require` marker. The PR is merged.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+                for param in settings:
+                    self.verify_change(node, param, settings[param], mark, expected_result)
+
+    @pytest.mark.require("scylladb/scylla-enterprise#1789")
```
Please, remove this `require` marker. The issue is closed.

andrzej-jackowski-scylladb replied:
Removed.

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+            with self.assert_entries_were_added(session, expected_new_entries, merge_duplicate_rows=False):
+                session.execute(auditted_query)
+
+    #@pytest.mark.require("scylladb/scylladb#22973")
```
Please, remove the comment.

andrzej-jackowski-scylladb commented on test/cluster/dtest/audit_test.py:
```diff
@@ -0,0 +1,1408 @@
+                table="cf",
+            )
+
+    @pytest.mark.require("scylladb/scylla-enterprise#3590")
```
Removed.

andrzej-jackowski-scylladb commented on test/cluster/dtest/audit_test.py:
```diff
@@ -696,7 +691,7 @@ def test_audit_type_invalid(self):
         expected_error = r"Startup failed: audit::audit_exception \(Bad configuration: invalid 'audit': invalid\)"
         self.ignore_log_patterns.append(expected_error)
-        self.cluster.nodes["node1"].watch_log_for(expected_error)
+        self.cluster.nodelist()[0].watch_log_for(expected_error)
```
Ok, reverted to nodes["node1"]

andrzej-jackowski-scylladb commented on test/cluster/dtest/audit_test.py:
```diff
@@ -741,7 +736,8 @@ def test_audit_categories_invalid(self, helper_class):
             expected_error = r"Startup failed: audit::audit_exception \(Bad configuration: invalid 'audit_categories': INVALID\)"
             self.ignore_log_patterns.append(expected_error)
-            self.cluster.nodes["node1"].watch_log_for(expected_error)
+            self.cluster.nodelist()[0].watch_log_for(expected_error)
```
Ok, reverted to nodes["node1"]

andrzej-jackowski-scylladb commented on test/cluster/dtest/tools/data.py:
```diff
@@ -0,0 +1,46 @@
+#
+import datetime
+import logging
+
```
Fixed

andrzej-jackowski-scylladb commented on test/cluster/dtest/tools/data.py:
```diff
@@ -0,0 +1,46 @@
+
+logger = logging.getLogger(__name__)
+
+
```
fixed

enaydanov commented:
@andrzej-jackowski-scylladb can you add a comment to this PR with the number of tests and the total execution time added?

andrzej-jackowski-scylladb commented:
> @andrzej-jackowski-scylladb can you add a comment to this PR with the number of tests and the total execution time added?

```
time pytest --test-py-init --mode=dev test/cluster/dtest/audit_test.py
...
collected 55 items                                                                                                                                                                                                
...
test/cluster/dtest/audit_test.py .x.x.x.x.x............x.x.x.......x....x.x.x...xx..xx..
...
=== 39 passed, 16 xfailed, 11 warnings in 240.76s (0:04:00) 

real	4m2.175s
user	0m48.507s
sys	0m28.077s
```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | sif |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)<br> :small_blue_diamond: cluster/dtest/audit_test | :white_check_mark: | spider1.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)<br> :small_blue_diamond: cluster/dtest/audit_test | :x: | spider5.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)<br> :small_blue_diamond: cluster/dtest/audit_test | :x: | spider9.cloudius-systems.com |

#### Build Details:
- Duration: 5 hr 27 min

yaronkaikov commented:
> > > #### 🔴 CI State: FAILURE
> > > Mode	Stage	Status	Node
> > > [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	sif
> > > dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider1.cloudius-systems.com
> > > [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider1.cloudius-systems.com
> > > [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)
> > > 🔹 cluster/dtest/audit_test	✅	spider1.cloudius-systems.com
> > > release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider5.cloudius-systems.com
> > > [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider5.cloudius-systems.com
> > > [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)
> > > 🔹 cluster/dtest/audit_test	❌	spider5.cloudius-systems.com
> > > debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider9.cloudius-systems.com
> > > [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)	✅	spider9.cloudius-systems.com
> > > [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17536/)
> > > 🔹 cluster/dtest/audit_test	❌	spider9.cloudius-systems.com
> > > #### Build Details:
> > > 
> > > * Duration: 5 hr 27 min
> > 
> > 
> > @enaydanov, my CI failed because custom executions of Unit Tests in debug and release failed. Just as other dtests moved to test.py, `audit test` is set as `run_in_dev` in `test/cluster/suite.yaml`.
> > Do you know why CI for other tests (e.g. #24230) passed with success, and mine is failing?
> 
> @yaronkaikov, we suspect that the new pipeline is responsible for the problem. It seems CI tries to run Unit Tests Custom in release and debug, and it fails because the new test is marked only as `run_in_dev`. Could you please take a look?

is there any way to know that in advance?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17552/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17552/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17552/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/dtest/audit_test
#### Failed Tests (1/2750):
- [test_permissions[AuditBackendTable].dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17552/testReport/junit/cluster.dtest.audit_test/TestCQLAudit/test_permissions_AuditBackendTable__dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_permissions[AuditBackendTable].dev.3)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider1.cloudius-systems.com

enaydanov commented:
> > @yaronkaikov, we suspect that the new pipeline is responsible for the problem. It seems CI tries to run Unit Tests Custom in release and debug, and it fails because the new test is marked only as `run_in_dev`. Could you please take a look?
> 
> is there any way to know that in advance?

There is `--list` option for `./test.py`:

```sh
$ ./test.py --list --mode=debug cluster/dtest
Test cluster/dtest not found
$ echo $?
1
$ 
```

So, you can run `./test.py --list` first for `Unit Test Custom` for a mode first.

andrzej-jackowski-scylladb commented:
v5:
 - Added missing wait for `read_barrier` in `test_permissions`
 - Added mechanism for filtering out `AUTH` and `USE ...` audit logs, that can appear at many points of tests due to driver behavior

@nuivall please review those recent changes

enaydanov commented on test/cluster/dtest/audit_test.py:
```diff
@@ -1016,6 +1011,10 @@ def test_insert_failure_doesnt_report_success(self):
 
         audit_paritition_nodes, insert_node, node_to_stop, query_to_fail = self._test_insert_failure_doesnt_report_success_assign_nodes(session=session)
 
+        #TODO remove when host_id is set on startup, we make it only to cache host_id
+        for node in audit_paritition_nodes + [insert_node] + [node_to_stop]:
+            node.cluster.manager.get_host_id(node.server_id) 
+
```
The test framework asks a node for host_id via ScyllaDB's REST API (`/storage_service/hostid/local` endpoint). So, it's not possible if the node is down. Probably, we can add parsing of scylla.log as a fallback option.

andrzej-jackowski-scylladb replied:
Reported https://github.com/scylladb/scylladb/issues/24473, will mention it in the comment

andrzej-jackowski-scylladb commented:
v6:
 - Removerted RF change in `test_parallel_syslog_audit` as it was no longer needed
 - Reported https://github.com/scylladb/scylladb/issues/24473, and mentioned it in the related comment

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17634/)
:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17634/)
#### Build Details:
- Duration: 11 min
- Builder: spider7.cloudius-systems.com

andrzej-jackowski-scylladb commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17634/) ❌ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17634/)
> 
> #### Build Details:
> * Duration: 11 min
> * Builder: spider7.cloudius-systems.com

The test modifies only test (python and yaml) files, but compilation failed with dwarf issue.
```
09:54:52  FAILED: build/debug/scylla.stripped 
09:54:52  scripts/strip.sh build/debug/scylla
09:54:52  ../../gdb/dwarf2/cooked-index.c:396: internal-error: finalize: Assertion `entry->canonical == nullptr' failed.
09:54:52  A problem internal to GDB has been detected,
09:54:52  further debugging may prove unreliable.
09:54:52  ----- Backtrace -----
09:54:52  0x564bb0faf5f7 ???
09:54:52  0x564bb13eacbc ???
09:54:52  0x564bb13eb08e ???
09:54:52  0x564bb161e5a4 ???
09:54:52  0x564bb1063e63 ???
09:54:52  0x564bb10654a8 ???
09:54:52  0x564bb1065671 ???
09:54:52  0x7f218905133a ???
09:54:52  0x7f21890513a8 ???
09:54:52  0x564bb162d21a ???
09:54:52  0x7f218932e523 ???
09:54:52  0x7f218904bfa7 ???
09:54:52  0x7f21890cffcb ???
09:54:52  0xffffffffffffffff ???
09:54:52  ---------------------
09:54:52  ../../gdb/dwarf2/cooked-index.c:396: internal-error: finalize: Assertion `entry->canonical == nullptr' failed.
09:54:52  A problem internal to GDB has been detected,
```

andrzej-jackowski-scylladb commented:
@piodul CI passed, please consider merging

piodul commented:
Queued
